"virt" machine command-line configurability

BrunoASMauricio opened the pull request:
Create a new, abstract virtual machine to centralize command line option handling.

Change the "virt" machine from being a "pure machine" to inherit from this new abstract machine.

Add ram-start as an option to the parent machine, and make "virt" use it during initialization.

These changes come about from the request made in https://github.com/foss-for-synopsys-dwc-arc-processors/qemu/issues/171 which requires us to either add support for configuring machines, or permanently change the ram_start address value. The flexibility of the former made it the chosen solution.

To change any machine option, instead of running with the usual: `-machine virt`/`-M virt`, use something like `-machine virt,ram-start=0`/`-M virt,ram-start=0`

Note:
The default value for "ram-start"  is the previous constant (0x80000000) so as not to disrupt currently automated usages of qemu (i.e. testsuite configurations)

 --  edit --

Note2:
The reason the virt machine was adapted but not arc-sim is because:
1) Unlike arc-sim, virt has ram starting at not 0, which means it is the one issue https://github.com/foss-for-synopsys-dwc-arc-processors/qemu/issues/171 is referring to (as currently we only have 2 machines)
2) For a reason that currently eludes me, arc-sim and virt have different "ram-start" values so I cannot pick a single default for both

evgeniy-paltsev commented:
I'm wondering if this require changes for current QEMU usage in zephyr, or this doesn't affect existing board (machine)?

Currently we run QEMU for HS with following command line:

```
qemu-system-arc -cpu archs -m 8M -nographic -no-reboot -monitor none -global cpu.firq=false -global cpu.num-irqlevels=15 -global cpu.num-irq=25 -global cpu.ext-irq=20 -global cpu.freq_hz=10000000 -global cpu.timer0=true -global cpu.timer1=true -global cpu.has-mpu=true -global cpu.mpu-numreg=16 -net none -pidfile qemu.pid -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -icount shift=6,align=off,sleep=off -rtc clock=vm -kernel zephyr.elf
```

Do we need to change someting in it?

BrunoASMauricio commented:
Hello @evgeniy-paltsev! As per the note in the PR description, this PR adds configurability for the "virt" machine, but only for ram-start (more can be easily added as needed) and the default value is the same as it was when it was hardcoded.

Therefore, you shouldn't need to change anything!

However, it would be best if you could checkout this branch and validate this for yourself since I do not have the binaries you use!

P.S.
I added another note to the description regarding why I changed virt but not arc-sim

BrunoASMauricio commented on hw/arc/virt.c:
```diff
@@ -119,7 +119,7 @@ static void virt_init(MachineState *machine)
     int n;
 
-    boot_info.ram_start = VIRT_RAM_BASE;
+    boot_info.ram_start = vms->ram_start;
```
You are very much correct, I forgot to save that change. Thank you for your comment!

abrodkin commented on hw/arc/arc_virt.c:
```diff
@@ -27,6 +27,7 @@
  * their name, parent structure, name inside parent structure, and default value
  */
 static Property arc_virt_properties[] = {
+    DEFINE_PROP_UINT64("ram-start", ARCVirtMachineState, ram_start, VIRT_RAM_BASE),
```
Since we use that value in that one place only I'd suggest to use explicit value just here, so it's immediately obvious what that value is, rather than having a definition in another header file. I.e. just use:
```c
DEFINE_PROP_UINT64("ram-start", ARCVirtMachineState, ram_start, 0x80000000)
```
And add in-place comment like that: for compatibility reason's assume default memory start at 0x8000_0000 (2 GiB), if needed change with command line option "-M virt,ram-start=0x12345678".

abrodkin commented on include/hw/arc/virt.h:
```diff
@@ -24,6 +24,8 @@
 #include "qom/object.h"
 #include "hw/qdev-properties.h"
 
+#define VIRT_RAM_BASE      0x80000000
```
See above.

abrodkin reviewed (changes requested):
I'd elaborate a bit more in the commit message on what do you do and why in that change. I.e. copy here a part of the PR message. Also see below another comment.
